pyro/infer/mcmc/util.py

```diff
@@ -0,0 +1,137 @@
+        # To be populated using the model trace once.
+        self._log_probs = {}
+        self._children = defaultdict(list)
+        self._enum_dims = {}
+        self._iarange_dims = {}
+        self._parse_model_structure(model_trace)
+
+    def _parse_model_structure(self, model_trace):
+        if not self.has_enumerable_sites:
+            return
+        if self.max_iarange_nesting == float("inf"):
+            raise ValueError("Finite value required for `max_iarange_nesting` when model "
+                             "has discrete (enumerable) sites.")
+        self._compute_log_prob_terms(model_trace)
+        # 1. Infer model structure - compute parent-child relationship.
+        sorted_ordinals = sorted(self._log_probs.keys())
+        for i in range(len(sorted_ordinals)):
+            child_node = sorted_ordinals[i]
+            for j in range(i-1, -1, -1):
+                cur_node = sorted_ordinals[j]
+                if cur_node < child_node:
+                    self._children[cur_node].append(child_node)
+                    break  # at most 1 parent.
+        # 2. Populate `iarange_dims` and `enum_dims` to be evaluated/
+        #    enumerated out at each ordinal.
+        self._populate_cache(self.root, frozenset(), set())
+
+    def _populate_cache(self, ordinal, parent_ordinal, parent_enum_dims):
+        """
+        For each ordinal, populate the `iarange` and `enum` dims to be
+        evaluated or enumerated out.
+        """
+        log_prob = self._log_probs[ordinal]
+        iarange_dims = sorted([frame.dim for frame in ordinal - parent_ordinal])
+        enum_dims = set((i for i in range(-log_prob.dim(), -self.max_iarange_nesting)
+                         if log_prob.shape[i] > 1))
+        self._iarange_dims[ordinal] = iarange_dims
+        self._enum_dims[ordinal] = sorted(list(enum_dims - parent_enum_dims))
+        for c in self._children[ordinal]:
+            self._populate_cache(c, ordinal, enum_dims)
+
+    def _compute_log_prob_terms(self, model_trace):
+        """
+        Computes the conditional probabilities for each of the sites
+        in the model trace, and stores the result in `self._log_probs`.
+        """
+        model_trace.compute_log_prob()
+        ordering = {}
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                if len(site["cond_indep_stack"]) == 0:
+                    ordering[name] = self.root
+                else:
+                    ordering[name] = frozenset(tuple(self.root) + site["cond_indep_stack"])
+
+        # Collect log prob terms per independence context.
+        log_probs = defaultdict(list)
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                if is_validation_enabled():
+                    check_site_shape(site, self.max_iarange_nesting)
+                log_probs[ordering[name]].append(site["log_prob"])
+
+        for ordinal in log_probs:
```

Comment:
nit: If you use `.items()` you can avoid half of the hashing of frozensets.
```py
for ordinal, terms in log_probs.items():
    self._log_probs[ordinal] = sum(terms)
```
Also, cool trick to avoid `0.0 + x` tensor ops!